internal/crypto/points.go

```diff
@@ -0,0 +1,143 @@
+package crypto
+
+import (
+	"crypto/elliptic"
+	"crypto/rand"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/zeebo/blake3"
+)
+
+/*
+High level api for operating on elliptic curve Points.
+*/
+
+// InitCurve instantiate an elliptic curve object given the curveName and returns the number of bytes
+// needed to encode a point on the curve
+func InitCurve() (curve elliptic.Curve, encodeLen int) {
+	curve = elliptic.P256()
+	encodeLen = len(elliptic.Marshal(curve, curve.Params().Gx, curve.Params().Gy))
+	return
+}
+
+// Points represents a point on an elliptic curve
+type Points struct {
+	curve elliptic.Curve
+	x     *big.Int
+	y     *big.Int
+}
+
+// NewPoints returns a blank points on an elliptic curve
+func NewPoints(curve elliptic.Curve) Points {
+	return Points{curve: curve, x: new(big.Int), y: new(big.Int)}
+}
+
+func newPoints(curve elliptic.Curve, x, y *big.Int) Points {
+	return Points{curve: curve, x: x, y: y}
+}
+
+// SetX sets the x coordiante of a point on elliptic curve
+func (p Points) SetX(newX *big.Int) Points {
+	p.x.Set(newX)
+	return p
+}
+
+// SetY sets the y coordinate of a point on an elliptic curve
+func (p Points) SetY(newY *big.Int) Points {
+	p.y.Set(newY)
+	return p
+}
+
+// Marshal converts a Points to a byte slice representation
+func (p Points) Marshal() []byte {
+	return elliptic.Marshal(p.curve, p.x, p.y)
+}
+
+// Unmarshal takes in a marshaledPoint byte slice and extracts the Points object
+func (p Points) Unmarshal(marshaledPoint []byte) error {
```

Comment:
For example, this is especially confusing. One would assume that since the point is passed by value, the original point won't be modified but we do it secretly by calling Set on the bignums.